graphql_compiler/compiler/emit_sql.py

```diff
@@ -88,27 +100,187 @@ def _create_query(node, context):
+    return _expression_to_sql(expression, node, context)
+
+
+def _expression_to_sql(expression, node, context):
+    """Recursively transform a Filter block predicate to its SQLAlchemy expression representation.
+
+    Args:
+        expression: expression, the compiler expression to transform.
+        node: SqlNode, the SqlNode the expression applies to.
+        context: CompilationContext, global compilation state and metadata.
+
+    Returns:
+        Expression, SQLAlchemy Expression equivalent to the passed compiler expression.
+    """
+    _expression_transformers = {
+        expressions.LocalField: _transform_local_field_to_expression,
+        expressions.Variable: _transform_variable_to_expression,
+        expressions.Literal: _transform_literal_to_expression,
+        expressions.BinaryComposition: _transform_binary_composition_to_expression,
+    }
+    expression_type = type(expression)
+    if expression_type not in _expression_transformers:
+        raise NotImplementedError(
+            u'Unsupported compiler expression "{}" of type "{}" cannot be converted to SQL '
+            u'expression.'.format(expression, type(expression)))
+    return _expression_transformers[expression_type](expression, node, context)
+
+
+def _transform_binary_composition_to_expression(expression, node, context):
+    """Transform a BinaryComposition compiler expression into a SQLAlchemy expression.
+
+    Recursively calls _expression_to_sql to convert its left and right sub-expressions.
+
+    Args:
+        expression: expression, BinaryComposition compiler expression.
+        node: SqlNode, the SqlNode the expression applies to.
+        context: CompilationContext, global compilation state and metadata.
+
+    Returns:
+        Expression, SQLAlchemy expression.
+    """
+    if expression.operator not in constants.SUPPORTED_OPERATORS:
+        raise NotImplementedError(
+            u'Filter operation "{}" is not supported by the SQL backend.'.format(
+                expression.operator))
+    sql_operator = constants.SUPPORTED_OPERATORS[expression.operator]
+    left = _expression_to_sql(expression.left, node, context)
+    right = _expression_to_sql(expression.right, node, context)
+    if sql_operator.cardinality == constants.CARDINALITY_UNARY:
+        left, right = _get_column_and_bindparam(left, right, sql_operator)
+        clause = getattr(left, sql_operator.name)(right)
+        return clause
+    elif sql_operator.cardinality == constants.CARDINALITY_BINARY:
+        clause = getattr(sql_expressions, sql_operator.name)(left, right)
+        return clause
+    elif sql_operator.cardinality == constants.CARDINALITY_LIST_VALUED:
+        left, right = _get_column_and_bindparam(left, right, sql_operator)
+        # ensure that SQLAlchemy treats the right bind parameter as list valued
+        right.expanding = True
+        clause = getattr(left, sql_operator.name)(right)
+        return clause
+    else:
```

Comment:
nit: At one point there was a subtle bug in the compiler where an if-elif tree like this one, where each branch was meant to return, actually had a code path that didn't return. With the `raise AssertionError` in the `else` branch, that meant that no error was raised (since the `else` wasn't taken, one of other the branches was taken but didn't return), and instead the function simply returned `None` to the caller ... which promptly exploded elsewhere.

So if all branches are meant to `return`, I'd pull the `AssertionError` out of the else, and make it say `Unreachable state reached` (+ as much contextual info as possible) instead of necessarily blaming the cardinality.

Reply:
That makes sense.  I remember that error, it is very hard to track down.